File: src/main/java/com/socialthingy/plusf/util/Try.java
package com.socialthingy.plusf.util;

import java.util.function.Consumer;

public class Try<T> {
    private Exception exception;
    private T success;

    public static <S> Try<S> success(final S success) {
        final Try<S> s = new Try<>();
        s.success = success;
        return s;
    }

    public static <S> Try<S> failure(final Exception exception) {
        final Try<S> s = new Try<>();
        s.exception = exception;
        return s;
    }

    private Try() {}

    public void ifSuccess(final Consumer<T> fn) {
        if (success != null) {
            fn.accept(success);
        }
    }

    public void ifFailure(final Consumer<Exception> fn) {
        if (exception != null) {
            fn.accept(exception);
        }
    }
}

File: plus-f/src/main/kotlin/com/socialthingy/plusf/wos/Try.kt
package com.socialthingy.plusf.wos

fun <T> Try(expr: () -> T): Try<T> {
    return try {
        Success(expr())
    } catch (ex: Exception) {
        Failure(ex)
    }
}

sealed class Try<out T> {
    abstract fun <U> map(fn: (T) -> U): Try<U>
    abstract fun ifDefined(fn: (T) -> Unit)
    abstract fun get(): T
}

class Success<out T>(val t: T): Try<T>() {
    override fun <U> map(fn: (T) -> U): Try<U> {
        return Success(fn(t))
    }

    override fun ifDefined(fn: (T) -> Unit) {
        fn(t)
    }

    override fun get(): T {
        return t
    }
}

class Failure<out T>(val ex: Exception): Try<T>() {
    override fun <U> map(fn: (T) -> U): Try<U> {
        return Failure(ex)
    }

    override fun ifDefined(fn: (T) -> Unit) {}

    override fun get(): T {
        throw ex
    }
}